[Kyh196201] 7주차 레스토랑 로그인 페이지 및 리뷰 작성 페이지 구현하기

kyh196201 opened the pull request:
트레이너님 안녕하세요 😀
이번 한 주도 잘부탁드립니다!

강의 영상이 길어서 첫 번째 영상만 보고 우선 과제를 진행했습니다.

---

현재까지 진행 상황입니다.

- [x] 로그인 페이지 라우터 구현
- [x] LoginForm 컴포넌트 및 테스트 구현
- [x] LoginFormContainer 컴포넌트 및 테스트 구현
- [x] LoginPage 컴포넌트 및 테스트 구현
- [x] ReviewForm 컴포넌트 및 테스트 구현
- [x] RestaurantDetail 테스트 추가
- [x] RestaurantContainer 테스트 추가
- [x] 리뷰 작성 api 구현
- [x] Reviews 컴포넌트 구현 및 리뷰 목록 보여주기
- [x] 로그인 api 에러 핸들링
- [x] 로그인 이후의 동작

---

질문이 있습니다.

1. 로그인 페이지에서 사용하는 컴포넌트들의 테스트에서 공통적으로 `getLabelText`를 이용해서 엘리먼트를 검색하고 있는데, 이때 인자로 전달하는 `E-mail, Password` 값을 공통으로 관리하고 싶습니다.
이때도, `fixtures` 하위에 파일 생성후에 관리하면 될까요??

2. LoginForm 컴포넌트 테스트

email, password props에 빈 문자열, `undefined`가 들어가는 두 가지 케이스를 테스트하고 싶어서 이렇게 작성했는데, 괜찮을까요..?

코드는 [이 페이지](https://stackoverflow.com/questions/45713938/jest-looping-through-dynamic-test-cases)을 참고했습니다.

```javascript
context('without email and password values', () => {
	const loginFields = [{
		email: undefined,
		password: undefined,
	}, {
		email: '',
		password: '',
	}];

	loginFields.forEach(({ email, password }) => {
		it('it renders empty email and password', () => {
			const { getByLabelText } = renderLoginForm({
				email,
				password,
			});

			expect(getByLabelText('E-mail')).toHaveValue('');
			expect(getByLabelText('Password')).toHaveValue('');
		});
	});
});
```

kyh196201 committed:
HomePage 테스트 수정

- Login 링크가 렌더링되었는지 확인하는 테스트를 추가했습니다.

kyh196201 committed:
HomePage 테스트 통과

- 로그인 페이지로 이동하기위한 Link를 추가했습니다.

kyh196201 committed:
LoginPage 테스트 추가

- 렌더링 시 페이지 제목을 렌더링했는지 확인하는 테스트를 작성했습니다.

kyh196201 committed:
LoginPage 테스트 통과

kyh196201 committed:
App 테스트 추가

- /login 경로로 접근했을 때에 대한 테스트를 추가했습니다.

kyh196201 committed:
App 테스트 통과

- /login path에 해당하는 라우터를 설정했습니다.

kyh196201 committed:
LoginPage 위치 수정

- pages 디렉토리 하위로 login 디렉토리를 추가했습니다.

kyh196201 committed:
LoginFormContainer 테스트 추가

- 각 input이 렌더링되었는지 확인하는 테스트를 추가했습니다.

kyh196201 committed:
LoginFormContainer 테스트 수정

- submit button을 렌더링했는지 확인하는 테스트를 추가했습니다.

kyh196201 committed:
LoginFormContainer 테스트 통과

kyh196201 committed:
LoginPage 테스트 추가

- 페이지 내부에 인풋과 버튼이 렌더링되었는지 확인하는 테스트를 추가했습니다.

kyh196201 committed:
LoginPage 테스트 통과

kyh196201 committed:
LoginForm 컴포넌트, 테스트 추가

kyh196201 committed:
LoginFormContainer 리팩토링

- LoginFormContainer 내부의 UI에 관련된 로직을 LoginForm 컴포넌트를 이용해서 분리했습니다.

kyh196201 committed:
Reducer 테스트 수정

- initialState에 loginFields의 초기값이 있는지 확인하도록 테스트를 수정했습니다.

kyh196201 committed:
Reducer 테스트 통과

- 초기 상태에 loginFields를 추가했습니다.

kyh196201 committed:
LoginForm 테스트 추가

- email, password props에 따라서 렌더링이 달라지는 것에대한 테스트를 추가했습니다.

kyh196201 committed:
LoginForm 테스트 통과

- email, password props 구현

kyh196201 committed:
LoginForm 테스트 추가

- email, password input를 변경하는 이벤트가 발생했을 때에 대한 테스트를 추가했습니다.

kyh196201 committed:
LoginForm 테스트 통과

- onChange를 props로 전달받고, 내부에서 handleChange를 구현했습니다.

kyh196201 committed:
로그인 테스트용 fixture 추가

kyh196201 committed:
LoginForm 테스트 제목 수정

- 이메일, 패스워드 값에 따라 테스트하는 context의 제목을 수정했습니다.

kyh196201 committed:
LoginFormContainer 테스트 수정

- 스토어에있는 email, password를 LoginForm에 전달하는지 확인하는 테스트를 추가했습니다.

kyh196201 committed:
LoginFormContainer 테스트 통과

kyh196201 committed:
LoginPage 테스트 수정

- LoginFormContainer에서 useSelector를 사용해서 발생하는 에러를 수정했습니다.

kyh196201 committed:
App 테스트 수정

- LoginFormContainer에서 useSelector를 사용해서 발생하는 에러를 수정하기위해서
useSelector에 loginFields를 추가했습니다.

daadaadaah commented on src/pages/login/LoginPage.test.jsx:
```diff
@@ -10,4 +10,19 @@ describe('<LoginPage />', () => {
+    const { getByText } = renderLoginPage();
+
+    expect(getByText('Log In', {
+      selector: 'button',
```
LoginPage의 `<h1>Log In</h1>`와 구별을 주기위해서 추가하신 요소인가용?ㅎㅎ

kyh196201 replied:
네 맞습니다. `getByText('Log In')`으로 검색했을 때, 페이지 상단에 있는 h1태그가 먼저 검색이되어서 버튼 여부와 상관없이 테스트가 계속 통과했습니다.
그래서 `Log In`이라는 텍스트를 가진 버튼을 찾는 방법이 필요했고, 구글링해서 해당 옵션을 찾아서 적용했습니다!

daadaadaah replied:
아항! 그럴 경우 `getByRole`로도 표현해주기도 합니당!ㅎㅎ
```jsx
getByRole('button', { name: 'Log In' })
```

kyh196201 replied:
감사합니다!!
알려주신 방법이 버튼을 찾는다는걸 더 명시적으로 표현하는 것 같습니다 🙌

daadaadaah commented on src/components/login/LoginForm.test.jsx:
```diff
@@ -20,4 +25,40 @@ describe('<LoginForm />', () => {
+    it('it renders current email and password', () => {
+      const { getByLabelText } = renderLoginForm(loginFields);
+
+      expect(getByLabelText('E-mail')).toHaveDisplayValue(loginFields.email);
+      expect(getByLabelText('Password')).toHaveValue(loginFields.password);
```
Email과 password 각각 다른 matcher 사용하셨는데, 혹시 특별한 이유가 있으실까용?ㅎㅎ

kyh196201 replied:
우선 email은 타입이 text여서 실제 화면에 입력한 값이 렌더링되는지 확인하기위해서 `toHaveDisplayValue`를 사용했습니다.

password의 경우에는 타입이 password여서 화면에 `...`형태로 보이기 때문에 실제 값을 가지고 있는지 테스트하면 좋을 것 같다고 생각해서 다른 `toHaveValue`를 사용했는데, 이럴 경우 같은 `matcher`로 통일하는게 좋을까요??

daadaadaah replied:
아니요~ 꼭 통일 안해주셔도 됩니다! 다른게 표현하신 이유가 그냥 궁금했어요!
그런데, 제가 테스트 해보니까 password 형태는 화면에 `...`형태로 보여서 `not.toHaveDisplayValue`으로 테스트가 통과할 거라고 예상했는데, 실제 `...`형태여도 `toHaveDisplayValue(loginFields.password)`으로 테스트가 통과해서, 값이 실제 화면 보이는 걸로(?) 판단하나 보네요. 그러면 `toHaveValue`로 통일해서 써도 괜찮을 것 같다는 생각이 드네용!  참고만 해주세요!ㅎㅎ

kyh196201 replied:
트레이너님 상세한 피드백 너무 감사합니다 :-)
막상 안되겠지하고 짐작만 했지, 실제로 테스트를 실행해보지는 않았네요 ..!!!
다음 PR부터는 실제 실행 결과까지 확인해보겠습니다 감사합니당 !!

daadaadaah commented on src/components/login/LoginForm.test.jsx:
```diff
@@ -20,4 +25,40 @@ describe('<LoginForm />', () => {
+  });
+
+  context('without email and password values', () => {
+    const loginFields = [{
+      email: undefined,
+      password: undefined,
+    }, {
+      email: '',
+      password: '',
+    }];
```
빈 문자열 정도만 테스트해도 괜찮을 것 같은데, email과 password가 undefined 값을 가질 때가 어느 상황을 생각하신 건가요?

kyh196201 replied:
`renderLoginForm`함수에 인자를 전달할 때 `default function parameter`를 사용해서 테스트 코드에 `undefined`에 대한 테스트를 추가했습니다.

생각해보니, 실제 컴포넌트에는 default 파라미터가 없는데, 테스트 코드에만 적용했었네요.. 
`renderLoginForm` 함수 수정하고 undefined에 대한 테스트도 제거했습니다 😀

daadaadaah commented on src/components/login/LoginForm.test.jsx:
```diff
@@ -0,0 +1,105 @@
+    });
+  });
+
+  describe('change email input control', () => {
+    it('calls handleChange', () => {
+      const { getByLabelText } = renderLoginForm();
+
+      expect(handleChange).not.toBeCalled();
+
+      fireEvent.change(getByLabelText('E-mail'), {
+        target: { value: 'tester@example.com' },
+      });
+
+      expect(handleChange).toBeCalledWith({
+        name: 'email',
+        value: 'tester@example.com',
+      });
+    });
+  });
+
+  describe('change password input control', () => {
+    it('calls handleChange', () => {
+      const { getByLabelText } = renderLoginForm();
+
+      expect(handleChange).not.toBeCalled();
+
+      fireEvent.change(getByLabelText('Password'), {
+        target: { value: 'test' },
+      });
+
+      expect(handleChange).toBeCalledWith({
+        name: 'password',
+        value: 'test',
+      });
+    });
+  });
```
이 2개의 테스트를 아래처럼 1개의 테스트로 묶어서 관리할 수 있어요! 한번 시도해보세요!
```js
  it('listens change events', () => {
  });
```

kyh196201 replied:
수정 완료했습니다! 감사합니다 😎

daadaadaah reviewed (commented):
오! 이전 기수분 PR 보면서 PR에 진행상황 남겨주신 점 너무 감사합니당!ㅎㅎ덕분에 리뷰하기 편했어요!ㅎㅎ승우님 PR이랑 commit 등을 보면 여러번 느끼는 거지만, 상대방 입장을 많이 생각하고 작업하시는 것 같아서 너무 좋습니당!👍 👍👍 지금 PR 에 나온 대로, 로그인 페이지 라우터 구현 빼고 현재 진행중이신거죵?ㅎㅎ 나머지도 다음 PR때 반영해봅시당👊


> 1. 로그인 페이지에서 사용하는 컴포넌트들의 테스트에서 공통적으로 getLabelText를 이용해서 엘리먼트를 검색하고 있는데, 이때 인자로 전달하는 E-mail, Password 값을 공통으로 관리하고 싶습니다.
이때도, fixtures 하위에 파일 생성후에 관리하면 될까요??

- 공통으로 관리하고 싶다는 의미가 어떤 의미일까용? 지금 [코드 작성해놓으신 것](https://github.com/CodeSoom/react-week7-assignment-1/pull/90/files#diff-bc6122ff39d356acbbd160a5f046b0fb76d2598e100b8ef282adf4576d8606dc)처럼 fixture에 넣어두고, 여러테스트에서 불러와 사용하고 싶다는 의미이실까용?

> 2. LoginForm 컴포넌트 테스트 - email, password props에 빈 문자열, undefined가 들어가는 두 가지 케이스를 테스트하고 싶어서 이렇게 작성했는데, 괜찮을까요..?
- 여러개 테스트 하고 싶을 때, 승우님이 작성하신대로 해도 괜찮은데, 저는 이 코드를 보고 `email과 password가 undefined 값을 가질 때가 어느 상황인 거지?`라는 생각이 들었습니당!ㅎㅎ[코멘트](https://github.com/CodeSoom/react-week7-assignment-1/pull/90#discussion_r925096592) 참고해주세요!

daadaadaah commented on src/pages/login/LoginPage.test.jsx:
```diff
@@ -0,0 +1,39 @@
+    },
+  }));
+
+  const renderLoginPage = () => render((<LoginPage />));
```
Page 컴포넌트를 테스트 할 때, 어떤 걸 감싸줘야 하던대, 무엇일까요? 🤔

kyh196201 replied:
`MemoryRouter`로 감싸서 렌더링하도록 테스트 코드 수정했습니다.

그런데 질문이 있습니다.

제가 강의에서 놓친걸수도 있는데, 렌더링하는 페이지 내부에서 `Link`와 같은 라우터 내부에서만 동작하는 요소가 있을 경우 `MemoryRouter`로 감싸서 테스트하는 것으로 알고 있는데, 혹시 `MemoryRouter`로 감싸야하는 다른 이유가 있나요??

kyh196201 committed:
LoginForm 테스트 수정

- renderLoginForm 함수의 default function parameter를 제거했습니다.
- email과 password가 undefined일 때에 대한 테스트를 제거했습니다.

kyh196201 committed:
LoginForm 테스트 수정

- change event에 대한 테스트 코드를 하나의 테스트로 합쳤습니다.

kyh196201 committed:
LoginPage 테스트 수정

- LoginPage 컴포넌트 렌더링 시 MemoryRouter로 래핑하도록 수정했습니다.

kyh196201 committed:
email, password input에 대한 fixture를 추가했습니다.

kyh196201 committed:
LoginFormContainer 테스트 추가

- email, password를 변경했을 때 setLoginFields을 dispatch하는지에 대한 테스트를 추가했습니다.

kyh196201 committed:
LoginFormContainer 태스트 통과

- setLoginFields 액션 creator를 추가했습니다.

kyh196201 committed:
LoginForm 테스트 제목 수정

kyh196201 committed:
reducer 테스트 추가

- setLoginFields 액션에 대한 테스트를 추가했습니다.

kyh196201 committed:
reducer 수정

- setLoginFields 액션에 대한 로직을 구현했습니다.

kyh196201 committed:
LoginForm 테스트 수정

- login 버튼 클릭에 대한 테스트를 추가했습니다.

kyh196201 committed:
LoginForm 테스트 통과

- handleSubmit props를 추가했습니다.

kyh196201 committed:
api 테스트 추가

- postLogin에 대한 테스트를 추가했습니다.

kyh196201 committed:
api 테스트 통과

- postLogin 함수를 구현했습니다.

kyh196201 committed:
reducer 테스트 수정

- initialState에 accessToken가 있는지 확인하도록 테스트를 수정했습니다.

kyh196201 committed:
reducer 테스트 추가

- initialState에 accessToken을 추가했습니다.

kyh196201 committed:
actions 테스트 수정

- 비동기 액션 requestLogin에 대한 테스트를 추가했습니다.

kyh196201 committed:
actions 테스트 통과

- 비동기 액션 postLogin과 액션 creator setAccessToken를 구현했습니다.

kyh196201 committed:
postLogin에 대한 mock 함수 추가

kyh196201 committed:
LoginFormContainer 테스트 수정

- login 버튼 클릭에 대한 테스트를 추가했습니다.

kyh196201 committed:
LoginFormContainer 테스트 통과

- login 버튼 클릭 시 requestLogin 액션을 이용해서 로그인 api를 요청하도록 구현했습니다.

kyh196201 commented:
트레이너님 피드백 주신 내용 반영하고, 로그인 API 요청까지 구현 완료했습니다 😀

로그인 이후 동작이랑 리뷰 작성 컴포넌트는 다음 PR에 구현해서 올리겠니다. 감사합니다 ㅎㅎ

---

> * 공통으로 관리하고 싶다는 의미가 어떤 의미일까용? 지금 [코드 작성해놓으신 것](https://github.com/CodeSoom/react-week7-assignment-1/pull/90/files#diff-bc6122ff39d356acbbd160a5f046b0fb76d2598e100b8ef282adf4576d8606dc)처럼 fixture에 넣어두고, 여러테스트에서 불러와 사용하고 싶다는 의미이실까용?

- 넵 맞습니다 ! 그래서 [`fixtures/login.js`](https://github.com/CodeSoom/react-week7-assignment-1/pull/90/files#diff-bc6122ff39d356acbbd160a5f046b0fb76d2598e100b8ef282adf4576d8606dc) 내부에 `EMAIL_INPUT`, `PASSWORD_INPUT`이라는 변수를 선언해서 테스트에 적용해봤는데 괜찮은 방법인지 궁금합니다 😀

kyh196201 committed:
reducer 테스트 추가

- setAccessToken 액션에 대한 테스트를 추가했습니다.

kyh196201 committed:
reducer 테스트 통과

- reducer에 setAccessToken 액션에 대한 로직을 구현했습니다.

daadaadaah commented on fixtures/login.js:
```diff
@@ -0,0 +1,15 @@
+
+export const PASSWORD = 'test';
+
+export const EMAIL_INPUT = {
+  value: 'tester@example.com',
+  label: 'E-mail',
+  name: 'email',
+};
+
+export const PASSWORD_INPUT = {
+  value: 'test',
+  label: 'Password',
+  name: 'password',
+};
```
value를 위에 EMAIL이랑 PASSWORD로 활용해도 해도 될 것 같아요!

kyh196201 replied:
> 음..일단 저는 처음 보는 형식이긴한대, 테스트 코드에서만 사용한다면 상관 없을 것 같아요!ㅎㅎ

피드백 감사합니다!
이 부분 적용하면서도 고민이 많았는데, 이번 과제에서 한번 사용해보겠습니다! 😀

daadaadaah reviewed (commented):
> 넵 맞습니다 ! 그래서 [fixtures/login.js](https://github.com/CodeSoom/react-week7-assignment-1/pull/90/files#diff-bc6122ff39d356acbbd160a5f046b0fb76d2598e100b8ef282adf4576d8606dc) 내부에 EMAIL_INPUT, PASSWORD_INPUT이라는 변수를 선언해서 테스트에 적용해봤는데 괜찮은 방법인지 궁금합니다 😀

- 음..일단 저는 처음 보는 형식이긴한대, 테스트 코드에서만 사용한다면 상관 없을 것 같아요!ㅎㅎ

kyh196201 committed:
RestaurantDetail 테스트 추가

- review form이 렌더링되었는지 확인하는 테스트를 추가했습니다.

kyh196201 committed:
RestaurantDetail 테스트 통과

- RestaurantDetail 내부에 review form을 구현했습니다.

kyh196201 committed:
ReviewForm 테스트 추가

- 인풋과 버튼이 렌더링되었는지 확인하는 테스트를 추가했습니다.

kyh196201 committed:
ReviewForm 테스트 통과

kyh196201 committed:
RestaurantDetail 수정

- review form UI에 대한 부분을 ReviewForm 컴포넌트로 대체했습니다.

kyh196201 committed:
RestaurantContainer 테스트 수정

- ReviewForm이 렌더링되었는지 확인하는 테스트를 추가했습니다.

kyh196201 committed:
Reducer 테스트 수정

- 초기 상태에 reviewFields 프로퍼티가 선언되어있는지 확인하도록 수정했습니다.

kyh196201 committed:
Reducer 테스트 통과

- initialState에 reviewFields를 추가했습니다.

kyh196201 committed:
ReviewForm 테스트 수정

- score, description props가 있을 경우, 없을 경우에 대한 테스트를 추가했습니다.

kyh196201 committed:
ReviewForm 테스트 통과

- score, description props를 구현했습니다.

kyh196201 committed:
ReviewForm 테스트 수정

- change 이벤트를 수신하는지에 대한 테스트를 추가했습니다.

kyh196201 committed:
ReviewForm 테스트 통과

- onChange props를 추가하고, 내부에 handleChange 함수를 구현했습니다.

kyh196201 committed:
RestaurantDetail 테스트 수정

- ReviewForm에서 발생한 change 이벤트를 수신하는지 확인하는 테스트를 추가했습니다.

kyh196201 committed:
RestaurantDetail 테스트 통과

- onChangeReviewForm props를 추가했고,
ReviewForm의 onChange props에 전달하도록 구현했습니다.

kyh196201 committed:
LoginForm 테스트 수정

- 컴포넌트 내부에 'Log In' 텍스트를 가진 엘리먼트가 더 없기때문에,
getByText의 selector 옵션을 제거했습니다.

kyh196201 committed:
LoginPage 테스트 수정

- 'Log In' 버튼을 찾는 함수를 getByRole로 변경했습니다.

kyh196201 committed:
login fixtures 수정

kyh196201 commented:
트레이너님 리뷰 요청드립니다!

이번 PR에 추가된 내용입니다.

- `ReviewForm` 컴포넌트 일부 구현, 테스트 작성
- `RestaurantDetail` 컴포넌트에 대한 테스트에 리뷰 관련 내용을 일부 추가했습니다.

감사합니다. 😀

daadaadaah commented on src/RestaurantDetail.jsx:
```diff
@@ -16,7 +16,9 @@ export default function RestaurantDetail({ restaurant }) {
       <MenuItems menuItems={menuItems} />
 
-      <ReviewForm />
+      <ReviewForm
+        onChange={onChangeReviewForm}
+      />
```
다른 속성들도 같이 적어주는게 좋을 같아요!

kyh196201 replied:
넵! 리뷰 작성 기능 구현하면서 나머지 속성도 추가했습니다 😎

daadaadaah reviewed (changes requested):
잘하셨어요! 계속 진행해주세요🚀

kyh196201 committed:
RestaurantDetail 테스트 수정

- reviewFields의 score, description 값이 있을 경우,
없을 경우에 대한 테스트를 추가했습니다.

kyh196201 committed:
RestaurantDetail 테스트 통과

- reviewFields를 props로 전달받도록 구현했습니다.

kyh196201 committed:
RestaurantDetail 테스트 수정

- reviewFields가 있는지 보다는 score, description 각 값이 있는지를 확인하는게
더 좋읗 것 같아서 테스트 제목을 수정했습니다.

kyh196201 committed:
review fields에 대한 fixture 추가

kyh196201 committed:
RestaurantContainer 테스트 수정

- score, description props가 있을 경우, 없을 경우에 대한 테스트를 추가했습니다.

kyh196201 committed:
RestaurantContainer 테스트 통과

- store에 있는 reviewFields를 RestaurantDetail에 전달하도록 수정했습니다.

kyh196201 committed:
RestaurantPage 테스트 수정

- 테스트 코드에 있는 useSelector의 state에 reviewFields가 없어서 에러가 발생했습니다.
그래서, reviewFields를 추가했습니다.

kyh196201 committed:
App 테스트 수정

- 테스트 코드에 있는 useSelector의 state에 reviewFields가 없어서 에러가 발생했습니다.
그래서, reviewFields를 추가했습니다.

kyh196201 committed:
setReviewFields 액션 creator 추가

kyh196201 committed:
RestaurantContainer 테스트 수정

- ReviewForm 내부의 인풋의 change 이벤트에 대한 테스트를 추가했습니다.

kyh196201 committed:
RestaurantContainer 테스트 통과

- setReviewFields 액션을 dispatch하는 handleChangeReviewForm를 구현하고,
RestaurantDetail 컴포넌트에 전달했습니다.

kyh196201 committed:
Reducer 테스트 수정

- setReviewFields에 대한 테스트를 추가했습니다.

kyh196201 committed:
Reducer 테스트 통과

- setReviewFields 액션에 대한 로직을 구현했습니다.

kyh196201 committed:
ReviewForm submit 이벤트 기능 구현 및 테스트 추가

kyh196201 committed:
RestaurantDetail submit 이벤트 기능 구현 및 테스트 추가

kyh196201 committed:
api 테스트 추가

- postReview api가 성공했을 때에 대한 테스트를 추가했습니다.

kyh196201 committed:
api 테스트 통과

- postReview api를 구현하고, mock api 함수를 추가했습니다.

kyh196201 committed:
RestaurantContainer 수정

- ReviewForm의 submit에 대한 테스트를 추가하고, handleSubmitReviewForm 함수를 구현했습니다.
- 우선 테스트를 통과시키기 위해서 handleSubmitReviewForm 내부에서 dispatch를 그냥 실행했습니다.

kyh196201 commented:
트레이너님 리뷰 요청드립니다!
피드백 주신 내용 반영하고, 리뷰 작성관련 추가 로직을 구현했습니다.

이번 PR에 추가된 내용입니다.

- `ReviewForm` 컴포넌트 구현 완료.
- `RestaurantDetail` 컴포넌트 리뷰 작성 로직 관련 테스트 추가 완료.
- `RestaurantContainer` 컴포넌트에 리뷰 작성 로직 관련 테스트 추가

---

그리고 질문이 있습니다. 

리뷰 작성 api 같은 경우는 api 요청에 성공했을 때 `response`가 따로 없을 것 같아서 `postReview` 함수의 리턴 값을 `boolean` 타입으로 설정했습니다.
트레이너님은 api 문서에 명시된 `response`가 없거나 `null`일 경우 반환 값을 어떻게 설정하시는지 궁금합니다 😀

daadaadaah commented on src/RestaurantDetail.jsx:
```diff
@@ -17,6 +18,8 @@ export default function RestaurantDetail({ restaurant, onChangeReviewForm }) {
       <MenuItems menuItems={menuItems} />
 
       <ReviewForm
+        score={score}
+        description={description}
```
`reviewFields`를 통으로 넘겨주는 것과 지금처럼 디스트럭처링해서 하나하나 넘겨주는 것이 어떤 차이점이 있을까요?ㅎㅎ
만약에, 속성의 이름이 변경되거나 속성이 추가되었을 때를 생각해보면 고민해보세요!

kyh196201 replied:
트레이너님 피드백 감사합니다 😊
이런 부분은 생각못했는데, 확실히 `reviewFields`를 통으로 넘겨주는 것이 더 좋을 것 같습니다!
왜냐하면, `reviewFields`의 속성이 바뀔경우 사용하는 컴포넌트에서만 대응해주면 되기 때문입니다 ㅎㅎ
이 부분도 리팩토링 진행할게요 감사합니다!

daadaadaah commented on src/RestaurantDetail.test.jsx:
```diff
@@ -63,27 +63,27 @@ describe('RestaurantDetail', () => {
   });
 
-  context('with review fields', () => {
+  context('with score and description', () => {
```
`reviewFields` -> `score and description`을 번경하셨는데,  속성이 늘어나거나 속성명이 바뀌다면, 테스트 설명도 수정해줘야 하는 번거로움이 있을 것 같은데, 어떻게 생각하시나요? 이제 테스트 코드 작성하시는거 익숙해졌으니, 테스트 코드도 유지보수 관점에서 어떻게 작성하면 좋을지 고민해봅시당!ㅎㅎ

kyh196201 replied:
테스트 코드의 유지 보수는 진짜 과제를 진행하면서 한번도 신경써보지 못했네요!
피드백 너무 감사합니다 🙌
이 부분도 실제 `reviewFields`를 사용하는 곳이 아니라 하위 컴포넌트에 넘겨주는 역할을 하므로 기존 제목이 더 좋을 것 같습니다.

daadaadaah commented on fixtures/review.js:
```diff
@@ -0,0 +1,3 @@
+export const SCORE = '5';
+
+export const DESCRIPTION = 'Good!';
```
`reviewfields`라는 객체로 관리해도 괜찮을 것 같아요!

daadaadaah commented on src/services/api.js:
```diff
@@ -27,3 +27,39 @@ export async function fetchRestaurant({ restaurantId }) {
+
+  await response.json();
+
+  return true;
```
백앤드 쪽에 문제가 생겨서 예기치 못하게 실패할 경우에도 true를 반환해주나용?ㅎㅎ

kyh196201 replied:
`try catch`를 이용해서 에러가 발생했을 경우 `false`를 반환하게하면 좋을 것 같습니다 감사합니다!!

daadaadaah replied:
음...API 응답에 대한 에러 헨들링(`try catch`)을 어디서 해줘야 되는지에 대해 먼저 생각해보시고, API 응답 성공/에러 여부에 따라 `true/false`를 반환하게 하는 방법이 괜찮을지에 대해 한번 고민해보세요!🤔

daadaadaah commented on src/actions.js:
```diff
@@ -85,3 +86,34 @@ export function loadRestaurant({ restaurantId }) {
+  };
+}
+
+export function requestLogin() {
+  return async (dispatch, getState) => {
+    const { loginFields } = getState();
+
+    const accessToken = await postLogin(loginFields);
+
+    dispatch(setAccessToken(accessToken));
+  };
+}
```
"API 응답 실패에 따른 예외 처리하기" 한번 도전해보시는건 어떠신가요?
try-catch 사용해서, API 응답을 실패했을 때 어떻게 처리할 것인지, 또한 그럴 때 사용자에게 어떤 UI를 보여줄지에 대해 한번 고민해보세용!

daadaadaah reviewed (commented):
> 리뷰 작성 api 같은 경우는 api 요청에 성공했을 때 response가 따로 없을 것 같아서 postReview 함수의 리턴 값을 boolean 타입으로 설정했습니다.
트레이너님은 api 문서에 명시된 response가 없거나 null일 경우 반환 값을 어떻게 설정하시는지 궁금합니다 😀

- api의 경우 백앤드 개발자랑 상의하고 정해서, 저만의 방식은 따로 생각해본적은 없네용😅 그냥 아예 response를 없게 한 적도 있고, 팀에서 응답 코드의 형태를 정해서 한적도 있고~ 지금 과제에서는 api를 수정할 수 없으니, 그냥 response가 없게 해도 괜찮을 것 같아요!😃

kyh196201 committed:
Reducer 테스트 추가

- clearReviewFields 액션에 대한 테스트를 추가했습니다.

kyh196201 committed:
actions 수정

- clearReviewFields 액션 creator를 구현했습니다.

kyh196201 committed:
Reducer 테스트 통과

- clearReviewFields 액션에 대한 로직을 구현했습니다.

kyh196201 committed:
actions 테스트 추가

- addReview 비동기 액션에 대한 테스트를 추가했습니다.

kyh196201 committed:
actions 테스트 통과

addReview 비동기 액션을 구현했습니다.
리뷰 작성이 성공하면, reviewFields를 비우고, 다시 레스토랑 정보를 조회하도록 구현했습니다.

kyh196201 committed:
RestaurantContainer 테스트 통과

- RestaurantDetail로 부터 onSubmitReviewForm 이벤트를 수신하면,
addReview 액션을 dispatch하도록 구현했습니다.

kyh196201 committed:
Reviews 테스트 추가

kyh196201 committed:
Reviews 테스트 통과

kyh196201 committed:
restaurant fixture 수정

- 리뷰 목록을 추가했습니다.

kyh196201 committed:
RestaurantDetail 테스트 수정

- 리뷰 목록이 렌더링되었는지 확인하는 테스트를 추가했습니다.
- 파일 내부에서 선언한 restaurant 변수를 사용하지 않고, fixture에 있는 restaurant를 사용했습니다.

kyh196201 committed:
RestaurantDetail 테스트 통과

- 리뷰 목록을 보여주는 Reviews 컴포넌트를 추가했습니다.

kyh196201 committed:
ReviewForm 테스트 수정

- ReviewForm이 props로 reviewFields을 전달받도록 테스트를 수정했습니다.

kyh196201 committed:
REVIEW_FIELDS fixture 추가

kyh196201 committed:
ReviewForm 테스트 통과

- reviewFields props를 받도록 수정했습니다.

kyh196201 committed:
RestaurantDetail 수정

- ReviewForm에 reviewFields를 그대로 전달하도록 수정했습니다.

kyh196201 committed:
RestaurantDetail 테스트 수정

kyh196201 commented:
트레이너님 피드백 주신 내용 반영하고, 리뷰 작성 및 목록 조회 기능 구현 완료했습니다.

남은 시간 동안에는 위에서 말씀해주신 api 에러 핸들링 관련 로직 및 스토리지 기능 추가하겠습니다.

감사합니다. 😊

daadaadaah commented on fixtures/restaurant.js:
```diff
@@ -5,4 +5,13 @@ export default {
   menuItems: [
     { id: 1, name: '떡볶이' },
   ],
+  reviews: [
+    {
+      id: 1,
+      restaurantId: 1,
+      name: '테스터',
+      score: 5,
+      description: 'Good!',
+    },
+  ],
```
배열인 경우, 저는 데이터를 2 ~ 3개 정도 추가해서 사용해요~!

kyh196201 replied:
피드백 감사합니다! 다른 케이스도 추가로 적용해서 테스트에 추가할게요 :-)

daadaadaah commented on src/RestaurantContainer.jsx:
```diff
@@ -18,6 +20,15 @@ export default function RestaurantContainer({ restaurantId }) {
+    dispatch(setReviewFields({ name, value }));
+  }
+
+  async function handleSubmitReviewForm() {
+    await dispatch(addReview({ restaurantId }));
+  }
```
`await/async` 로 해주신 특별한 이유가 있으신가용?

kyh196201 replied:
리뷰를 작성하고나서 다시 레스토랑 정보를 조회하는 로직을 이 곳에 작성하려다가 다시 지웠는데, 그때 작성한 코드가 남아있는 것 같습니다. 수정하겠습니다 🙌

daadaadaah commented on src/actions.test.js:
```diff
@@ -100,4 +104,43 @@ describe('actions', () => {
+      });
+    });
+
+    it('calls clearReviewFields', async () => {
+      await store.dispatch(addReview({ restaurantId: 1 }));
+
+      const actions = store.getActions();
+
+      expect(actions[0]).toEqual(clearReviewFields());
+      expect(actions[1]).toEqual(setRestaurant(null));
+      expect(actions[2]).toEqual(setRestaurant({}));
+    });
```
1. 테스트 설명을 `dispatch XXX`로 하는게 좀더 적절할 것 같아요!
2. setRestaurant가 2번 호출되는 건가용?

kyh196201 replied:
1. 앗 그렇네요! 액션을 호출하는 것이니까 dispatch가 더 좋을 것 같습니다. 감사합니다!
2. 네 맞습니다. 리뷰 작성에 성공하면 레스토랑 정보를 다시 조회하게 되는데, `loadRestaurant` 비동기 액션 내부에서 `setRestaurant`이 두번 실행되어서 이 부분에 대한 테스트를 작성했습니다.
이 테스트를 작성할 때가 좀 어려웠는데 어떤것을 테스트해야할지 잘 모르겠더라구요 ㅠㅠ
`addReview` 함수 안에있는 api를 모킹한 다음에 이 api가 실행되었는지 테스트해도 괜찮을까요??

daadaadaah replied:
아니요! 테스트 잘 작성하셨어요!ㅎㅎ addReview 전후로 발생하니까 2번 해줘야 되는거 맞네용!!👍

daadaadaah reviewed (commented):
마지막 PR 화이팅하세용!💪

kyh196201 commented:
트레이너님 이번 한주도 피드백 너무 감사합니다 :-)
제가 주말동안 과제를 거의 진행못했는데, 지금 추가 기능 구현하려고 합니다.
피드백 주신 내용도 반영하고 api 실패에 대한 기능도 구현해보겠습니다. 
다음 과제는 벌써 마지막인데 더 열심히해볼게요 ! 감사합니다 😃

kyh196201 committed:
LoginFormContainer 테스트 수정

- 로그인 했을 경우에 대한 테스트를 추가했습니다.

kyh196201 committed:
LoginFormContainer 테스트 통과

- 로그인되어있을 경우 Log out 버튼을 렌더링하도록 수정했습니다.

kyh196201 committed:
로그인 이후의 동작 구현

- 로그인 후 clearLoginFields 액션을 dispatch해서 loginFields을 초기화하는 로직을 구현했습니다.

kyh196201 committed:
api 테스트 수정

- postLogin api 실패 케이스에 대한 테스트 추가

kyh196201 committed:
api 테스트 통과

- postLogin에서 에러가 발생했을 경우에 대한 로직을 추가했습니다.

kyh196201 committed:
setError 액션 및 로직 추가

kyh196201 committed:
postLogin mocking 함수 변경

- 실패 케이스에 대한 테스트를 위해서 jest.fn()으로 변경했습니다.

kyh196201 committed:
actions 테스트 수정

- requestLogin 액션 실패 케이스에 대한 로직을 추가했습니다.

kyh196201 committed:
LoginForm 에러 메시지 구현

kyh196201 committed:
LoginFormContainer 로그인 실패 에러에 대한 로직 추가

kyh196201 committed:
App, LoginPage 테스트 오류 수정

- LoginFormContainer에서 스토어에 있는 errors를 사용하기 때문에 발생하는 오류를 수정하기 위해서
App, LoginPage의 테스트에 있는 useSelector에 errors를 추가했습니다.

kyh196201 committed:
LoginFormContainer 로그아웃 기능 구현

kyh196201 committed:
e2e 테스트 수정

kyh196201 committed:
restaurant fixture 수정

- 리뷰 목록에 항목을 추가했습니다.

daadaadaah commented on src/containers/login/LoginFormContainer.test.jsx:
```diff
@@ -0,0 +1,161 @@
+
+  const renderLoginFormContainer = () => render((<LoginFormContainer />));
+
+  context('not logged in', () => {
```
`context`는 with 또는 when을 활용해서 테스트 이름을 작성해주는게 좋아요!
`with/without login` 또는 `with/without accessToken` 로 표현해주셔도 좋을 것 같아요!

kyh196201 replied:
트레이너님 피드백 감사합니다 !!
이런 부분은 습관화해야겠어요 ㅎㅎ 감사합니다

daadaadaah commented on src/reducer.test.js:
```diff
@@ -232,4 +236,18 @@ describe('reducer', () => {
   });
+
+  describe('setError', () => {
+    it('update errors', () => {
```
사소한 것이지만, 계속 it문을 `changes XXX`로 했으니, 여기도 맞춰줍시당!ㅎㅎ

kyh196201 replied:
제목 수정했습니다 감사합니다!

daadaadaah commented on src/actions.test.js:
```diff
@@ -107,20 +110,50 @@ describe('actions', () => {
-      store = mockStore({
-        email: '',
-        password: '',
+    context('with valid email and password', () => {
```
`with valid loginFields`로 했을 때와 지금의 코드와 어떤 점이 다른지 한번 고민해보세요!ㅎㅎ

kyh196201 replied:
`reviewFields`와 마찬가지로 내부 속성이 변경될 경우에 테스트 제목을 수정해야겠네요!
유지보수에 안좋을 것 같습니다 😂

daadaadaah commented on src/containers/login/LoginFormContainer.jsx:
```diff
@@ -24,9 +25,16 @@ export default function LoginFormContainer() {
   };
 
+  const handleClickLogout = () => {
+    dispatch(setAccessToken(''));
```
API는 없지만, `requestLogin`처럼 따로 `requestLogout`액션 만들어줘서 거기서 로그아웃에 필요한 로직(accessToken 초기화 + localstorage에서 accessToken 제거)을 관리해주시는 것도 좋을 것 같아요!

kyh196201 replied:
넵 이 부분 수정했습니다.
`redux thunk`를 이용한 비동기 액션은 무조건 안에 비동기적인 로직이 있어야되는줄 알았는데, 여러 액션을 처리하는 용도로 사용할 수도 있더라구요! 
피드백 감사합니다!

daadaadaah commented on src/services/api.test.js:
```diff
@@ -67,4 +70,47 @@ describe('api', () => {
+      });
+    });
+
+    it('returns accessToken', async () => {
+      const accessToken = await postLogin({ email: '', password: '' });
+
+      expect(accessToken).toEqual('ACCESS_TOKEN');
+    });
+
+    it('should fail', async () => {
+      global.fetch = jest.fn().mockResolvedValueOnce({
+        ok: false,
+      });
+
+      await expect(async () => {
+        await postLogin({ email: '', password: '' });
+      }).rejects.toThrow('로그인에 실패했습니다.');
+    });
```
서로 대비되는 상황이니, context로 구분해줘주셔도 괜찮을 것 같아요!

kyh196201 replied:
피드백 감사합니다! 수정했습니다 😀

daadaadaah reviewed (commented):
한주동안 수고 많으셨습니당!! 주간회고를 통해 한주를 되돌아보는 시간을 가져보세요~
남은 마지막 8주차 과제도 화이팅하시구요~ 💪

kyh196201 committed:
LoginFormContainer 테스트 수정

- 로그인 했을 경우, 안 했을 경우에 대한 테스트의 제목을 변경했습니다.

kyh196201 committed:
Reducer 테스트 수정

setError 테스트 코드의 제목을 수정했습니다.
파일 내부에서 제목의 통일성을 위해서 changes errors로 변경했습니다.

kyh196201 committed:
actions 테스트 수정

requestLogin 테스트에서 context의 제목을 수정했습니다.
loginFields의 속성이 변경될 경우 테스트 제목을 유지보수하기 힘들기 때문에,
각 속성의 이름보다 loginFields을 직접 명시했습니다.

kyh196201 committed:
requestLogout 액션 구현

kyh196201 committed:
LoginFormContainer 수정

로그아웃 시 requestLogout 액션을 dispatch하도록 수정했습니다.

kyh196201 committed:
api 테스트 수정

postLogin 테스트 코드를 context를 이용해서 분리했습니다.

kyh196201 committed:
global.localStorage 모킹

kyh196201 committed:
storage 유틸 함수 테스트 및 구현

kyh196201 committed:
requestLogin 액션 테스트 수정

로그인 시 로컬스토리지에 토큰을 저장하는지에 대한 테스트를 작성했습니다.

kyh196201 committed:
App 테스트 수정

처음 앱에 진입했을 때 로컬스토리지에 저장된 accessToken이 있으면 setAccessToken 액션을
실행하는지에대한 테스트를 작성했습니다.

kyh196201 committed:
App 테스트 추가

- 로컬스토리지에 accessToken이 없을 경우에 대한 테스트를 추가했습니다.

kyh196201 commented:
트레이너님 이미 시간이 지났는데 피드백 해주셔서 너무 감사합니다😎
스토리지 구현, e2e 테스트까지 모두 완료했습니다.
